<a href="https://colab.research.google.com/github/kaicque/Projeto-Final-Bootcamp-DS/blob/main/Notebooks/Tratamento_dados.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Notebook de tratamento de dados pré analise

## Importação das bibliotecas utilizadas no notebook:

In [ ]:
# Importando Pandas para atribuir o dataset a uma variável.
import pandas as pd

# Importando o numpy.
import numpy as np

# Importando o LabelEncoder para categorizar os dados da dezena de idade
from sklearn.preprocessing import LabelEncoder

# Retirada dos alertas
import warnings
warnings.filterwarnings('ignore')

## Criação de funções:

In [ ]:
# Com essa função apliquei o metodo 'back' e 'foward' fill nas colunas de valores continuos, retornando ao final o dataframe completo e aproveito para transformar a coluna "AGE_PERCENTIL" em dados categoricos.
def preenche_tabela_frente(dados):
    features_continuas_colunas = dados.iloc[:,13:-2].columns
    features_continuas = dados.groupby("PATIENT_VISIT_IDENTIFIER", as_index=False)[features_continuas_colunas].fillna(method='bfill').fillna(method='ffill')
    features_categoricas = dados.iloc[:,:13]
    saida = dados.iloc[:,-2:]
    dados_finais = pd.concat([features_categoricas,features_continuas,saida], ignore_index=True,axis=1)
    dados_finais.columns = dados.columns
    # Tratando os dados de dezena de idade em categorias
    set(dados_finais['AGE_PERCENTIL'])
    label_encoder = LabelEncoder()
    dados_finais['AGE_PERCENTIL'] = label_encoder.fit_transform(dados_finais['AGE_PERCENTIL'])
    return dados_finais

In [ ]:
# Função para filtrar a primeira hora de todos os paciente e assinalando que ele pode ter ido pra UTI em seguida.
def prepare_window(rows):
    if(np.any(rows["ICU"])):
        rows.loc[rows["WINDOW"]=="0-2", "ICU"] = 1
    return rows.loc[rows["WINDOW"] == "0-2"]

In [ ]:
# Função para verificar quantidade de dados nulos a base possue
def verifica_nulos(dados):
  dados_faltantes = pd.DataFrame()
  dados_faltantes['nulos']  = data=dados.isnull().sum()
  dados_faltantes['perc_dados_faltantes'] = (dados_faltantes["nulos"]/dados.shape[0])*100
  dados_faltantes = dados_faltantes.sort_values('perc_dados_faltantes',ascending=False)
  return dados_faltantes

In [ ]:
#Função para remoção de colunas com multicolinearidade
def remove_corr_var(dados,valor_corte):
  
  matrix_corr = dados.iloc[:,4:-2].corr().abs()
  matriz_upper = matrix_corr.where(np.triu(np.ones(matrix_corr.shape),k=1).astype(np.bool))
  excluir = [coluna for coluna in matriz_upper if any(matriz_upper[coluna] > valor_corte)]

  return dados.drop(excluir,axis=1)

## Importação dos dados originais:

In [ ]:
# Importando a planilha original do Kaggle, criando e visualizando o dataframe.
dados = pd.read_excel('https://github.com/kaicque/Projeto-Final-Bootcamp-DS/raw/main/Dados/Original/Kaggle_Sirio_Libanes_ICU_Prediction.xlsx')
# Exibindo os tipos de dados existentes e quantidade de entradas
dados.info()
# Exibindo os primeiros dados
dados.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1925 entries, 0 to 1924
Columns: 231 entries, PATIENT_VISIT_IDENTIFIER to ICU
dtypes: float64(225), int64(4), object(2)
memory usage: 3.4+ MB


PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...    WINDOW  ICU
0                         0            1  ...       0-2    0
1                         0            1  ...       2-4    0
2                         0            1  ...       4-6    0
3                         0            1  ...      6-12    0
4                         0            1  ...  ABOVE_12    1

[5 rows x 231 columns]

Logo na exibição dos primeiros dados, verificamos alguns valores nulos. Então já deixamos o alerta ligado para verificar quantos dados que faltam.

Para isso criaremos um novo dataframe para ter uma ideia de qual o % dos dados estão nulos

In [ ]:
#Criação do data framme dados faltantes para verificar a quantidade de colunas que tem pelo menos 30% de dados totais faltantes(Nulos/NA).
dados_faltantes = verifica_nulos(dados)
dados_faltantes.query('perc_dados_faltantes > 30')

nulos  perc_dados_faltantes
P02_VENOUS_MIN                   1104             57.350649
LEUKOCYTES_MAX                   1104             57.350649
PC02_VENOUS_MEAN                 1104             57.350649
PC02_VENOUS_MIN                  1104             57.350649
PC02_VENOUS_MAX                  1104             57.350649
...                               ...                   ...
BLOODPRESSURE_DIASTOLIC_MEDIAN    685             35.584416
HEART_RATE_MEAN                   685             35.584416
BLOODPRESSURE_SISTOLIC_MEAN       685             35.584416
BLOODPRESSURE_DIASTOLIC_MEAN      685             35.584416
BLOODPRESSURE_DIASTOLIC_MIN       685             35.584416

[216 rows x 2 columns]

Com essa tabela verificamos que temos 216 colunas com pelo menos 30% de dados nulos, e o maior percentual fica na casa dos 57%. 

Então precisaremos de uma ação para tratamento desses dados.

## Tratamento dos dados:

### Preenchiento dos dados faltantes:

Verificando a documentação do desafio, os dados faltantes era um problema conhecido e fica explicitado na seguinte frase que pode ser encontrada no [Kaggle](https://www.kaggle.com/S%C3%ADrio-Libanes/covid19) :

* "***It is reasonable to assume that a patient who does not have a measurement recorded in a time window is clinically stable, potentially presenting vital signs and blood labs similar to neighboring windows.***"

Logo podemos preencher os dados faltantes com os dados de janelas anteriores ou posteriores, o que já é uma sugestão do desafio. 

Por tanto, imaginando que o dado vazio supõe que não houve alteração da medida anterior, adotei como premissa preencher os dados vazios com informações da entrada anterior e, quando a condição não se aplicar, puxando dado posteriores. 

In [ ]:
#Criando um novo data frame com os dados preenchidos 
dados_preenchidos = preenche_tabela_frente(dados)
dados_preenchidos.describe()

PATIENT_VISIT_IDENTIFIER  ...          ICU
count               1925.000000  ...  1925.000000
mean                 192.000000  ...     0.267532
std                  111.168431  ...     0.442787
min                    0.000000  ...     0.000000
25%                   96.000000  ...     0.000000
50%                  192.000000  ...     0.000000
75%                  288.000000  ...     1.000000
max                  384.000000  ...     1.000000

[8 rows x 230 columns]

In [ ]:
# Verificando se o novo data frame ainda contém algum dado nulo
dados_faltantes = verifica_nulos(dados_preenchidos)
dados_faltantes.query('perc_dados_faltantes > 30')

Empty DataFrame
Columns: [nulos, perc_dados_faltantes]
Index: []

Com os dados preenchidos, verificamos que temos 1925 linhas de dados de pacientes com 384 pacientes diferentes.

### Filtro dos dados utilizáveis:

Com o intuito de obter a resposta o mais breve possível para melhorar a tomada de decisão precisamos filtrar os dados em 2 etapas:
  * Tirar os paciêntes que foram para a UTI na primeira janela
  * E nos demais paciêntes, demarcar logo na primeira janela se ele foi para a UTI em janelas futuras e utilizar somente ela.

In [ ]:
# Primeira etapa: tirar os pacientes que foram internados logo na primeira hora
a_remover = dados_preenchidos.query("WINDOW=='0-2' and ICU==1")['PATIENT_VISIT_IDENTIFIER'].values
dados_preenchidos = dados_preenchidos.query("PATIENT_VISIT_IDENTIFIER not in @a_remover")
#dados_preenchidos = dados_preenchidos.dropna()
dados_preenchidos.describe()

PATIENT_VISIT_IDENTIFIER  ...          ICU
count               1760.000000  ...  1760.000000
mean                 192.818182  ...     0.201705
std                  110.637724  ...     0.401387
min                    0.000000  ...     0.000000
25%                   97.750000  ...     0.000000
50%                  191.500000  ...     0.000000
75%                  289.250000  ...     0.000000
max                  384.000000  ...     1.000000

[8 rows x 230 columns]

Logo com apenas a primeira etapa concluida, verificamos a quantidade de dados diminuir pelo filtro de pacientes que foram pra UTI na primeira janela.

In [ ]:

# Segunda etapa: Filtrar apenas a primeira janela de todos os paciêntes e colocar se foram pra UTI ou não em qualquer janela do tempo.
dados_preenchidos = dados_preenchidos.groupby("PATIENT_VISIT_IDENTIFIER").apply(prepare_window)
dados_preenchidos.describe()

PATIENT_VISIT_IDENTIFIER  ...         ICU
count                352.000000  ...  352.000000
mean                 192.818182  ...    0.463068
std                  110.763735  ...    0.499344
min                    0.000000  ...    0.000000
25%                   97.750000  ...    0.000000
50%                  191.500000  ...    0.000000
75%                  289.250000  ...    1.000000
max                  384.000000  ...    1.000000

[8 rows x 230 columns]

Com a segunda etapa concluida, podemos observar uma quantidade menor de dados, que corresponde a apenas 1 linha por paciente. Essa abordagem foi utilizada para tentar antecipar o conhecimento de quais paciêntes podem ir pra UTI logo com o resultado do primeiro exame.

## Exclusão de colunas com multicolinearidade.

Nesta etapa, visamos reduzir redundâncias dos nossos futuros modelos de machine learning. Aplicando uma regua de corte de 0.95.

In [ ]:
# Criação do data frame sem as colunas com multicolinearidade
dados_limpos_sem_corr = remove_corr_var(dados_preenchidos,0.95)

In [ ]:
dados_limpos_sem_corr.describe()

PATIENT_VISIT_IDENTIFIER  AGE_ABOVE65  ...  HEART_RATE_DIFF_REL         ICU
count                352.000000   352.000000  ...           352.000000  352.000000
mean                 192.818182     0.457386  ...            -0.976421    0.463068
std                  110.763735     0.498890  ...             0.059606    0.499344
min                    0.000000     0.000000  ...            -1.000000    0.000000
25%                   97.750000     0.000000  ...            -1.000000    0.000000
50%                  191.500000     0.000000  ...            -1.000000    0.000000
75%                  289.250000     1.000000  ...            -1.000000    1.000000
max                  384.000000     1.000000  ...            -0.535878    1.000000

[8 rows x 99 columns]

Nesse ultimo tratamento, conseguimos visualizar uma redução na quantidade de colunas do dataframe, passando a ter 99 colunas.

## Salvando o data frame tratado para aplicação na analise exploratória e no modelo de machine learning

In [ ]:
dados_limpos_sem_corr.to_excel('dados_tratado.xlsx', index=False)